[1단계 - 자동차 경주 구현] 이프(송인봉) 미션 제출합니다.

sinb57 opened the pull request:
안녕하세요! 이프입니다~
프리코스 때부터 코드리뷰를 받고 싶어서 엄청 기대하고 있었는데, 막상 리뷰받으려니 많이 떨리네요ㅎ

나름대로 많은 생각을 가지고서 코드를 작성했지만, 어떻게 보이실지 모르겠습니다.ㅜ
가장 궁금한 질문을 먼저 드리고, 그 외 부분은 지적해주시는대로 답변드리겠습니다!

1. 구조 설계
입출력 로직과 비즈니스 로직을 명확히 분리하기 위해
Controller와 Service, View로 분리하여 구조를 설계했는데
이 방식이 괜찮은지, 아니라면 어떠한 문제점이 있는지 궁금합니다.

2. 자동차 전진 기능
자동차 전진 기능은 랜덤값에 따라 이뤄지는 로직입니다.
어떻게 하면 해당 기능을 테스트할 수 있을지 프리코스 때부터 고민했었고
당시 사용한 방법을 이번 과제에서도 사용을 했습니다.
NumberPicker 인터페이스를 생성해서 미리 지정한 숫자를 반환할 수 있는 CustomNumberPicker 구현체를 만듦으로써
전진 조건에 사용되는 랜덤값을 순차적으로 지정할 수 있도록 했습니다.
나름 괜찮은 방식이라고 생각하지만 보다 확실한 답을 듣고 싶어서 질문드립니다.
이 방식에 대한 럿고님의 생각이 궁금하며, 또 현업에서는 랜덤값에 대한 테스트를 어떻게 진행하는지도 궁금합니다.

sinb57 committed:
feat: 문자열 덧셈 계산기 기능 구현 (페어프로그래밍 종료)

sinb57 committed:
refactor: 테스트케이스 리펙토링

* refactor: 기존의 @Test을 @ParameterizedTest로 변경

* refactor: 테스트 메소드명 수정

hanull committed:
docs: 구현할 기능 목록 정의

hanull committed:
docs: 구현 목록 체크 박스 추가

hanull committed:
feat: 입력된 자동차 이름에 대한 예외처리 구현

hanull committed:
feat: 자동차 이름 입력 기능 구현

hanull committed:
feat: 이동횟수 입력 예외처리 구현

hanull committed:
feat: 자동차 전진 기능 구현

hanull committed:
feat: 우승자 선정 기능 구현

hanull committed:
feat: 출력 기능 및 프로그램 실행 로직 구현

hanull committed:
fix: 자동차 이름 입력이 쉼표(,)로 끝나는 경우 예외가 발생하지 않는 이슈 해결

hanull committed:
refactor: 프로그램 로직 메서드 분리 (페어프로그래밍 종료)

sinb57 committed:
refactor: Separator 클래스 생성을 통한 split 로직 리팩토링

* refactor: 명확한 의미 전달을 위해 Calculator 메소드명 변경

* feat: Separator 클래스 생성
  - 쉼표(,)와 콜론(:)을 이용한 split 로직 구현
  - 커스텀 구분자를 이용한 split 로직 구현

sinb57 committed:
refactor: 자동차 이름 검증 로직 클래스 분리

sinb57 committed:
refactor: 자동차 이름 입력 로직 리팩토링

* test: 자동차 이름 입력 기능 테스트 수정

* feat: Delimiter enum 클래스 생성

sinb57 committed:
refactor: Car 클래스 toString 리팩토링

* feat: Delimiter 클래스에 콜론(:), 바(-)에 대한 concat, repeat 메소드 구현

sinb57 committed:
refactor: 우승자 출력 메소드 리팩토링

* feat: Delimiter 클래스에 쉼표(,)에 대한 join 메소드 구현

sinb57 committed:
refactor: 이동횟수 숫자 검증 로직 리팩토링

* test: 이동횟수 입력 테스트 구현

* refactor: Round 클래스에서 이뤄지던 숫자 검증 로직을 InputView 클래스로 이동

* refactor: InputView 클래스에서 이동횟수 입력 메소드의 반환 값을 int형으로 변경

sinb57 committed:
refactor: 예외처리 클래스 세분화

sinb57 committed:
style: view 패키지 구조 변경 및 controller/picker 패키지 생성

sinb57 committed:
refactor: RoundRange enum 클래스 생성을 통한 Round 클래스 리팩토링

* style: domain 패키지 구조 세분화 진행

sinb57 committed:
refactor: RandomRange enum 클래스 생성을 통한 랜덤값 생성 로직 리팩토링

* feat: RandomRange enum 클래스 생성

sinb57 committed:
style: 컨벤션에 맞춰 라인 변경

sinb57 committed:
refactor: 출력문 상수 처리

sinb57 committed:
refactor: 예외처리 메세지 상수 처리

sinb57 committed:
refactor: AppConfig 클래스 분리

sinb57 committed:
refactor: Service 클래스 생성을 통한 Controller 클래스 로직 분리

* style: picker 패키지를 service 패키지로 이동

* refactor: Controller 클래스 내 출력 로직, 메소드 분리

* feat: GameService 클래스 생성

sinb57 committed:
refactor: CarNameValidator static 메소드 분리 및 접근제어자 변경

sinb57 committed:
refactor: Round 검증 로직 클래스 분리

* fix: RoundRange.MINIMUM 값 변경 (0->1)
  - 양수 검증 로직에 대한 테스트 실패 해결

* feat: RoundValidator 클래스 추가

sinb57 committed:
test: 라운드실행 및 우승자선정 테스트메소드에 대한 케이스 추가

sinb57 committed:
test: 모든 테스트 클래스에 @SuppressWarnings("NonAsciiCharacters") 추가

sinb57 committed:
refactor: 자동차 전진 로직 리팩토링

* refactor: CarCollection 클래스 내 전지 로직을 for-loop에서 stream으로 변경

* refactor: Car 클래스 내 이동 메소드 분리

* feat: ForwardCondition enum 클래스 생성

sinb57 committed:
refactor: 사용하지 않는 메소드 제거 및 줄 정리

* refactor: CustomNumberPicker 필드 변수명 변경

* style: enum 클래스에 대한 패키지 구조 변경

sinb57 committed:
refactor: Service 클래스에 Round 객체를 생성함에 따른 Controller 클래스 리팩토링

sinb57 committed:
refactor: 인터페이스 생성에 따른 클래스명 변경

sinb57 committed:
refactor: Controller의 run 메소드 로직을 application으로 이동

sinb57 committed:
refactor: 이동거리 비교 로직을 Car 클래스로 책임 전가

sinb57 committed:
Merge branch 'sinb57-racingcar' into sinb57

sinb57 committed:
refactor: 최대 이동거리 구하는 로직 리팩토링

sinb57 committed:
style: 라인 수정

ksy90101 commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	private List<String> extractCustomUnits(String targetValue) {
+		Matcher matcher = Pattern.compile(CUSTOM_FORMAT_REGEX).matcher(targetValue);
```
이펙티브 자바 아이템 6. 불필요한 객체를 만들지 마라를 참고하시면, Matcher는 비싼 객체로 만드는데 메모리를 많이 사용하는 객체입니다. 그럼 이런 객체는 어떻게 하면 좋을까요? 아래 글을 남겨둘테니, 한번 확인해보시고 수정해주시면 좋을거 같습니다 :) 

https://github.com/keesun/study/blob/master/effective-java/item6.md#%EB%AC%B4%EA%B1%B0%EC%9A%B4-%EA%B0%9D%EC%B2%B4

ksy90101 commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,41 @@
+	private List<String> extractCustomUnits(String targetValue) {
+		Matcher matcher = Pattern.compile(CUSTOM_FORMAT_REGEX).matcher(targetValue);
+		if (matcher.find()) {
+			return new ArrayList<>(Arrays.asList(matcher.group(1).split("")));
```
1과 ""을 상수로 빼 의미있는 이름을 지워주시면 더 가독성이 높아질듯 합니다 :) 
아울러 `matcher.group(1).split("")` 해당 값은 어떤 값을 의미하는지 처음 보는 개발자는 이해하기가 어려운데요.
변수로 추출해 의미있는 이름을 지어주시면 좋을거 같습니다 :)

ksy90101 commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,41 @@
+		if (matcher.find()) {
+			return new ArrayList<>(Arrays.asList(matcher.group(1).split("")));
+		}
+		return new ArrayList<>();
```
매번 비어있는 객체를 생성해서 사용한다면, 메모리상에 무리가 갈수 있습니다. Java API중에 Collections를 보시면 emptyList()가 있는데, 이걸 사용해 매번 생성하지 않도록 하면 어떨까요?

sinb57 replied:
로직을 변경해서 해당 문제를 해결했습니다!

ksy90101 commented on src/main/java/racingcar/AppConfig.java:
```diff
@@ -0,0 +1,40 @@
+	public final OutputView outputView;
+
+	private AppConfig() {
+		this.outputView = new OutputViewImpl();
+		this.reader = new ConsoleReader();
+		this.inputView = new InputViewImpl(reader);
+		this.numberPicker = new RandomNumberPicker();
+		this.gameService = new GameServiceImpl(numberPicker);
+		this.gameController = new GameControllerImpl(gameService, inputView, outputView);
```
싱글톤을 사용하실려고 하신거 같은데, 위와 같이 처리한다면, 변경에 유연하지 못할거 같습니다.
이걸 생성자에서 받아서, 처리하는 방식으로 변경하면 스프링처럼 의존성을 관리할수 있을거 같은데, 한번 시도 해보시면 어떨까요?

ksy90101 commented on src/main/java/racingcar/domain/car/condition/ForwardCondition.java:
```diff
@@ -0,0 +1,17 @@
+package racingcar.domain.car.condition;
+
+public enum ForwardCondition {
```
👍

ksy90101 commented on src/main/java/racingcar/domain/car/condition/ForwardCondition.java:
```diff
@@ -0,0 +1,17 @@
+
+public enum ForwardCondition {
+
+	MINIMUM(4);
```
이런식으로 구현했다면 최댓값도 validation 할수 있을거 같은데, 한번 추가해볼까요?

ksy90101 commented on src/main/java/racingcar/domain/car/condition/ForwardCondition.java:
```diff
@@ -0,0 +1,17 @@
+
+	MINIMUM(4);
+
+	private final int value;
```
value라는 단어는 굉장히 추상적인 단어라고 생각합니다.
좀더 의미있는 단어를 고민해보시면 어떨까요?
예를들면 ForwardNumber같은 단어가 있을거 같아요 :)

ksy90101 commented on src/main/java/racingcar/domain/car/condition/NameLength.java:
```diff
@@ -0,0 +1,17 @@
+
+	LENGTH(5);
+
+	private final int value;
```
여기도 동일합니다. value보단 더 의미있는 단어를 작성해주세요.
예를들면 nameLength같은 단어가 있을수 있겠네요.

ksy90101 commented on src/main/java/racingcar/domain/car/condition/NameLength.java:
```diff
@@ -0,0 +1,17 @@
+
+public enum NameLength {
+
+	LENGTH(5);
```
최소라는 단어를 추가하면 어떨까요?
나중에 최대 값이 정책으로 validation이 결정될수도 있으니깐요:)

ksy90101 commented on src/main/java/racingcar/domain/car/validator/CarNameValidator.java:
```diff
@@ -0,0 +1,47 @@
+		instance.validate(names);
+	}
+
+	void validate(List<String> names) {
```
default-private을 사용하신 이유가 있으실까요?

ksy90101 commented on src/main/java/racingcar/domain/car/validator/CarNameValidator.java:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	private boolean isNameDuplicated(List<String> names, String name) {
+		return (Collections.frequency(names, name) > 1);
```
매직넘버가 있네요 :)

sinb57 replied:
해당 부분을 고민하던 중, 요구사항이 변경되어 동일이름을 허용하되 개수 제한을 둘 수 있다는 생각을 하게 되었습니다.
그래서 NameDuplication 이라는 Enum 클래스를 별도로 생성해서, 이름 중복에 대한 책임을 전가했습니다.

ksy90101 commented on src/main/java/racingcar/exception/car/CarNameEmptyException.java:
```diff
@@ -0,0 +1,12 @@
+import racingcar.exception.RacingCarException;
+import racingcar.exception.car.message.CarNameExceptionMessage;
+
+public class CarNameEmptyException extends RacingCarException {
```
RacingCarEaxception으로 Exception을 하나로 묶어서 처리하는건 너무 좋네요 👍

ksy90101 commented on src/main/java/racingcar/view/input/InputView.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.List;
+
+public interface InputView {
```
View도 인터페이스로 분리하신 이유가 있으실까요?

ksy90101 commented on src/main/java/racingcar/view/input/InputViewImpl.java:
```diff
@@ -0,0 +1,45 @@
+	private List<String> parseNames(String inputValue) {
+		inputValue = appendSpaceBeforeSplit(inputValue);
+		return Arrays.stream(Delimiter.splitWithComma(inputValue))
+				.map(String::trim).collect(Collectors.toList());
```
```suggestion
				.map(String::trim)
				.collect(Collectors.toList());
```

ksy90101 commented on src/test/java/racingcar/domain/CarCollectionTest.java:
```diff
@@ -0,0 +1,128 @@
+public class CarCollectionTest {
+
+	private void exceptionTest(Class throwableClass, List<String> carNames) {
+		assertThrows(throwableClass, () -> new CarCollection(carNames)
+		);
```
```suggestion
		assertThrows(throwableClass, () -> new CarCollection(carNames));
```

ksy90101 reviewed (changes requested):
안녕하세요 :) 이프 첫번째 미션하느라 너무 고생 많으셨습니다 💯 
깔끔하게 잘 구현되어 있고 이름도 잘 지어주셔서 이해하는데 어렵지 않았던거 같아요.
간단한 피드백을 남겼으니, 확인해주시면 감사하겠습니다.
질문은 언제든지 댓글이나 DM으로 주시면 확인하고 답변 드리겟습니다!~

- 입출력 로직과 비즈니스 로직을 명확히 분리하기 위해
Controller와 Service, View로 분리하여 구조를 설계했는데
이 방식이 괜찮은지, 아니라면 어떠한 문제점이 있는지 궁금합니다.

제 의견은 일반적으로 가장 많이 사용하는 5layer 계층 방식이기 때문에, 현재 구현하신 방식에서는 좋게 구조를 설계했다고 생각이 듭니다. 아울러 리뷰에도 남겨두긴 했지만, Controller, View, Service를 모두 interface를 만들어서 작업하신 이유가 궁금하네요 :) 

- 자동차 전진 기능은 랜덤값에 따라 이뤄지는 로직입니다. 어떻게 하면 해당 기능을 테스트할 수 있을지 프리코스 때부터 고민했었고 당시 사용한 방법을 이번 과제에서도 사용을 했습니다. NumberPicker 인터페이스를 생성해서 미리 지정한 숫자를 반환할 수 있는 CustomNumberPicker 구현체를 만듦으로써 전진 조건에 사용되는 랜덤값을 순차적으로 지정할 수 있도록 했습니다. 나름 괜찮은 방식이라고 생각하지만 보다 확실한 답을 듣고 싶어서 질문드립니다.

지금 하신 방식이 현업에서도 테스트 하기 어려운 부분은 제어할수 잇도록 인터페이스를 분리하고 있습니다 :) 지금 하신 방식은 너무 좋은 방식이에요.
[인터페이스를 분리하여 테스트하기 좋은 메서드로 만들기](https://tecoble.techcourse.co.kr/post/2020-05-17-appropriate_method_for_test_by_interface/)
읽어 보셨는지 모르겠지만, 현재 이프가 하신 방식을 정리한 글로 한번 읽어보시면 좋을듯 합니다!

sinb57 commented:
상세한 리뷰 감사합니다!ㅎ 하나씩 읽어보고 질문에 대해 답변드리면서 피드백을 반영하도록 하겠습니다!

sinb57 commented:
Separator 클래스를 보시면 extractCustomUnits 메소드와 extractOriginalString 메소드가 유사한 형태를 지니고 있습니다.
이전에 `프리코스-자판기` 미션에서 도입해보았듯, `템플릿 콜백 패턴`을 적용하고 싶지만,
당시와는 다르게 Separator 클래스의 두 메소드는, return 값이 다르다는 점에서 도통 방법이 떠오르지 않네요.
해당 중복 형태는 분리가 불가능한 건가요?

프리코스-자판기 미션: https://github.com/sinb57/java-vendingmachine-precourse/blob/develop/src/main/java/vendingmachine/controller/MachineControllerImpl.java

sinb57 committed:
refactor: 자동차 이름 검증 로직 리팩토링

* refactor: CarNameValidator 클래스 리팩토링
 - 직관적인 의미 파악을 위해, 각각의 validate 메소드의 조건문을 별도의 메소드로 분리
 - 한줄로 사용된 stream 부분을 라인별로 분리

* feat: DuplicateName Enum 클래스 생성을 통해 이름 중복 확인의 책임을 전가

sinb57 committed:
refactor: RandomRange Enum 클래스 내 매직넘버 제거

sinb57 committed:
style: 컨벤션 변경에 따른 코드 수정 (tab -> 4space)

sinb57 committed:
refactor: 자동차 Dto 클래스 생성을 통해 도메인과 View 간의 관계 최소화

* refactor: CarCollection 클래스의 getStatuses 메소드 반환 타입 변경

* refactor: 자동차 도메인이 지닌 toString 메소드를 CarStatusDto 클래스로 이동

* feat: 자동차 도메인에 대한 CarStatusDto 클래스 생성

sinb57 committed:
refactor: Service 클래스의 playRound 메소드, 책임에 따른 메소드 분리

sinb57 committed:
refactor: Controller 클래스의 playGame 메소드, 통일성을 위해 메소드 분리

sinb57 committed:
refactor: CarCollection 클래스 내 메소드, 명확성을 위해 메소드 분리

* style: stream 코드 개행

* refactor: getFarthestLocation 메소드, 메소드 분리

sinb57 committed:
refactor: NAME_TOO_LONG_EXCEPTION_MESSAGE 메시지 포맷팅

* style: 변수 명명 및 stream 개행

sinb57 committed:
refactor: 자동차 이름 검증 로직 순서 변경

  - 길이 검증 이전에, Null 검증을 우선 시행하도록 배치

sinb57 committed:
refactor: 자동차 이름에 대한 검증 로직을 Car 클래스에서 행하도록 변경

* test: 자동차 생성에 대한 테스트 코드 작성

* feat: CarCollectionValidator 생성, 자동차 이름 중복에 대한 검증 로직 수행

* feat: CarValidator 생성, 자동차 이름에 대한 검증 로직 수행

sinb57 committed:
test: 테스트 코드 리팩토링

* test: 예외테스트 목적으로 추출된 exceptionTest 매개변수 Class 타입 명시

* test: @MethodSource 어노테이션에 대한 메소드 추출, 별도의 Provider 클래스 생성

* test: @DisplayName 어노테이션을 통한 테스트 설명 추가, 메소드명 변경

sinb57 committed:
refactor: 문자열 뎃셈 계산기 입력 검증 로직 분리

* style: 변수 명칭 수정 및 stream문 개행

* feat: 예외처리 상황에 따른 Exception 클래스 생성

sinb57 committed:
refactor: 문자열 분리 로직 리팩토링

* test: Calulator 문자열 덧셈 기능 테스트 케이스 추가

* test: Separator 문자열 분리 로직 기능 테스트 코드 생성

* refactor: extractCustomUnits, extractString 메소드 중복 코드 추출

* refactor: (SEPARATOR_FORMAT)정규식 표현 방식 수정을 통한, 문자열 분리 로직 리팩토링

sinb57 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,57 @@
+			validateNumberIsNegative(number);
+			return number;
+		} catch (NumberFormatException exception) {
+			throw new RuntimeException("숫자 이외의 값을 입력하셨습니다.");
```
요구사항을 다시 살펴보니 해당 부분은 `RuntimeException` 예외를 throw를 던지라고 명시가 되어 있네요.
때문에 `문자열 덧셈 계산기` 미션에 대해서는 `RuntimeException` 예외를 throw 하도록 놔두었습니다.

![image](https://user-images.githubusercontent.com/22176552/153839150-c085ce58-87a5-455e-bcb0-ee5878ed6fc2.png)

sinb57 committed:
style: CarCollection 클래스명을 CarGroup 으로 명칭 변경

sinb57 committed:
refactor: DI 관리 방식 리팩토링

sinb57 commented:
피드백 주신 내용을 반영해서 리팩토링을 완료하여, 다시 한번 PR 리뷰 요청드립니다!! 잘 부탁드립니다!ㅎ

sinb57 commented on src/main/java/racingcar/domain/car/CarCollection.java:
```diff
@@ -0,0 +1,44 @@
+import racingcar.domain.car.validator.CarNameValidator;
+import racingcar.service.picker.NumberPicker;
+
+public class CarCollection {
```
Collection이라는 이름을 함부로 쓰는 것이 마음에 걸려서, CarGroup 이라고 클래스명을 변경했습니다!

ksy90101 commented on src/main/java/racingcar/domain/car/CarGroup.java:
```diff
@@ -0,0 +1,50 @@
+import racingcar.domain.car.validator.CarGroupValidator;
+import racingcar.service.picker.NumberPicker;
+
+public class CarGroup {
```
이 부분은 개인적으로 컨벤션일꺼 같은데요. (컨벤션이라는것은 달라질수가 있습니다.)
저는 `Collection classes, such as vectors and hashes should always be pluralized. Alternately, collection classes can also be prefixed with the word some.` 이와 같은 컨벤션을 지켜서 사용하고 있어요. 
사실 이펙티브 자바는 좋은 책 이지만, 너무 거기에만 따라가면 안된다고 생각이 드는데요.
한번 이프도 여러가지 컨벤션을 찾아보시면 좋을거 같아요.
참고로 클린코드에서도 아래와 같이 설명이 되어 있습니다.
클린코드에서도 아래와 같이 설명을 하고 있어요.
```
개발자에게는 특수한 의미를 가지는 단어(List, Array등)는 실제 저수준 자료구조를 표현한 의미가 아닌 이상 accountList와 같이 변수명에 붙이지 말자. 고수준 비즈니스 정책에 맞는 복수형 단어를 사용합니다
```

제가 생각하는 컨벤션이란 저만 읽을수 있는 코드를 작성하는게 아니라 다른 사람도 읽을수 있는 코드를 작성하는것이라고 생각이 드는데요. 한번 다른 크루들에게도 이야기를 들어보고 이프만의 이유있는 컨벤션을 가져보세요 :)

ksy90101 commented on src/main/java/calculator/NumberValidator.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    private boolean isNumberNegative(int number) {
+        return (number < 0);
```
제 생각은 굳이? 라는 의미가 들어요.
괄호를 사용하지 않아도 해당 코드는 충분히 읽힐수 있다고 생각이 들어요.
조건을 분명히 할수 있다고 하셨지만, 한번 제거해보시고 조건을 분명히 할수 있는지 확인해보세요.
```
7.3 return Statements
A return statement with a value should not use parentheses unless they make the return value more obvious in some way. Example:

Copy
Copied to ClipboardError: Could not Copy
return;

return myDisk.size();

return (size ? size : defaultSize);
```

ksy90101 commented on src/main/java/racingcar/domain/car/condition/ForwardCondition.java:
```diff
@@ -0,0 +1,17 @@
+
+    MINIMUM(4);
+
+    private final int number;
```
과연 number도 좋은 의미일까요? 저는 개인적으로 value와 포괄적인 의미라고 생각이 드는데요.
enum에서 왜 여러개의 필드를 사용할수 있을까요?
제 의견은 하나의 상수에서 여러가지의 값을 가질수 있기 때문이라고 생각이 드는데요.
대표적인 예시로, Name이라는 클래스가 있을때, 과연 필드에 어떤게 있을까요?
이렇게 생각해보신다면 enum명과 필드명이 굳이 달라야 하는 이유가 있을까요?
한번 고민해보세요~

ksy90101 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private boolean isNullOrEmpty(String text) {
+        return (text == null || text.isEmpty());
```
혹시 String의 isEmpty() vs isBlank()의 차이를 아실까요?
현재 상황에서는 어떤 케이스가 어울리는지 한번 확인해보시고 답변 달아주시면 좋을거 같습니다.

ksy90101 replied:
저도 동일하게 생각을 하였습니다.
따라서 이와 같은 키워드를 드린거였구요 ㅎㅎ 잘 적용해주신거 같아서 감사합니다 :)

ksy90101 commented on src/test/java/racingcar/domain/CarGroupTest.java:
```diff
@@ -0,0 +1,104 @@
+    private static final String providerPath = "racingcar.domain.provider.CarGroupTestProvider#";
+
+    private void exceptionTest(Class<? extends RacingCarException> exceptionClass, List<String> carNames) {
+        assertThrows(exceptionClass, () -> new CarGroup(carNames)
+        );
```
```suggestion
        assertThrows(exceptionClass, () -> new CarGroup(carNames));
```

ksy90101 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,53 @@
+public class CarTest {
+
+    private void exceptionTest(Class<? extends RacingCarException> exceptionClass, String name) {
+        assertThrows(exceptionClass, () -> new Car(name)
+        );
```
```suggestion
        assertThrows(exceptionClass, () -> new Car(name));
```

ksy90101 commented on src/test/java/racingcar/view/input/reader/CustomReader.java:
```diff
@@ -0,0 +1,16 @@
+package racingcar.view.input.reader;
+
+public class CustomReader implements Reader {
```
👍

ksy90101 commented on src/test/java/racingcar/view/input/RoundInputTest.java:
```diff
@@ -0,0 +1,37 @@
+import racingcar.exception.round.RoundNotNumericException;
+import racingcar.view.input.reader.CustomReader;
+
+public class RoundInputTest {
```
Junit5부터는 아래와 같은 이유로 public을 사용하고 있지 않는데요. 한번 읽어보시고 수정해주세요.
```
Class and method visibility
Test classes, test methods, and lifecycle methods are not required to be public, but they must not be private.

It is generally recommended to omit the public modifier for test classes, test methods, and lifecycle methods unless there is a technical reason for doing so – for example, when a test class is extended by a test class in another package. Another technical reason for making classes and methods public is to simplify testing on the module path when using the Java Module System.
```

https://junit.org/junit5/docs/current/user-guide/#writing-tests-classes-and-methods

ksy90101 commented on src/main/java/racingcar/exception/car/message/CarNameExceptionMessage.java:
```diff
@@ -0,0 +1,13 @@
+
+public class CarNameExceptionMessage {
+
+    public static String NAME_DUPLICATED_EXCEPTION_MESSAGE = "자동차의 이름은 중복될 수 없습니다.";
+    public static String NAME_EMPTY_EXCEPTION_MESSAGE = "자동차의 이름은 공백이 될 수 없습니다.";
+    public static String NAME_NULL_EXCEPTION_MESSAGE = "자동차의 이름은 NULL이 될 수 없습니다.";
+    public static String NAME_TOO_LONG_EXCEPTION_MESSAGE =
+            String.format("자동차의 이름은 %d자 이하여야 합니다.", NameLength.getMaximum());
```
정적인 경우에는 모두 final을 붙여 불변을 유지해주시는게 좋습니다.
그 이유를 한번 찾아보시고, 답변을 남겨주시면 좋을거 같아요.

ksy90101 reviewed (changes requested):
안녕하세요 이프 :) 리뷰어 럿고입니다.
수정해주신 코드 및 남겨주신 답변들 잘 봤습니다.
좀 더 수정할 것들과, 조금더 고민해보면 좋을것들, 좀 더 찾아보고 학습해보면 좋을것들을 작성해서 피드백을 남겨두었습니다.
확인하시고 다시 리뷰 요청주시면 감사하겠습니다 :) 
질문이 있다면 언제든지 질문 해주세요~

sinb57 committed:
refactor: GameApplication DI 수정

sinb57 committed:
refactor: 예외메시지 상수 문자열의 접근제어자 변경

sinb57 committed:
test: 테스트 클래스 접근제어자 변경 및 import 순서 재정렬

sinb57 committed:
refactor: ExceptionMessage 클래스를 enum 으로 변경

sinb57 committed:
refactor: 이름 중복 허용 개수에 대한 enum 클래스, 상수 명칭 수정

sinb57 committed:
refactor: 이름 최대 길이 조건에 대한 enum 클래스, 상수 명칭 수정

sinb57 committed:
refactor: 실행 횟수 최소 조건에 대한 enum 클래스, 상수 명칭 수정

sinb57 committed:
refactor: 전진 조건 최소 범위에 대한 enum 클래스, 상수 명칭 수정

sinb57 committed:
refactor: 리스트 반환 타입의 메소드, 반환 값의 불변성 보장

sinb57 committed:
refactor: 구조 단순화, 필요하지 않은 interface 제거

sinb57 committed:
refactor: CarStatusDto 불변성 지니도록 수정

sinb57 committed:
refactor: 출력 문자열 상수 static final 지정

sinb57 committed:
refactor: 싱글톤 인스턴스 명칭 수정

sinb57 committed:
refactor: return 문에 사용된 조건식에 대해, 불필요한 괄호 제거

sinb57 committed:
refactor: Validator 클래스, private 생성자 생성 및 인스턴스 상수 명칭 수정

sinb57 committed:
refactor: CarGroup 일급컬렉션 이름 변경(Cars)

sinb57 committed:
refactor: 구조 단순화를 위한, 커스텀 Exception 제거

sinb57 committed:
refactor: 랜덤 숫자 범위에 대한 enum 클래스, 상수 명칭 수정

sinb57 committed:
refactor: validator 클래스, 검증 메소드 접근제어자 private 지정

sinb57 committed:
refactor: Delimiter 클래스, 필드 변수 명칭 수정

sinb57 committed:
refactor: 커스텀 Exception 에러 문자열 상수 명칭 수정

sinb57 committed:
refactor: 문자열 덧셈 계산기, 커스텀 Exception 제거

sinb57 committed:
refactor: 커스텀 Exception 클래스 구조 변경 및 불필요한 클래스 정리

sinb57 committed:
refactor: intellij-java-wooteco-style 컨벤션 적용

sinb57 commented on src/main/java/calculator/exception/CalculatorException.java:
```diff
@@ -0,0 +1,9 @@
+package calculator.exception;
+
+public class CalculatorException extends RuntimeException {
```
아래와 같은 이유로 Exception을 커스텀하는 것이 좋다고 판단했습니다.
1. 메세지를 따로 확인하지 않아도 클래스명으로 어떤 상황인지 유추할 수 있다는 점
2. 예외테스트 작성 같은 상황에서 확인하고자 하는 Exception을 특정 지정할 수 있다는 점
   (스프링의 컨트롤러에서도 Exception을 특정 짓는 경우가 있다고 합니다.)

하지만, 이번에 `럿고`의 피드백을 받고
에러 문자열을 관리하던 ExceptionMessage 클래스를 enum으로 변경하는 과정에서
Exception 클래스의 개수만큼이나 많은 코드를 수정해야만 했습니다.

이 과정에서 Exception 클래스가 제대로 설계된 것이 많는지 돌이켜 보게 되었고,
각각의 클래스가 예외 메세지(상황)에 종속적으로 생성되어 있음을 깨달았습니다.

이렇게 되면 예외 상황을 추가하거나 삭제, 변경할 때마다
부득이하게 클래스를 건드리게 되며,
이로 인해 연관된 코드들 또한 수정해야 하는 상황이 초래됩니다.

때문에 새롭게 PR을 요청하는 지금은, IllegalArgumentException 표준 예외을
상속받는 단 하나의 Exception 클래스를 생성해두었고
각 상황별 예외 메세지는 enum으로 분리해두었습니다.

sinb57 committed:
refactor: ExceptionStatus 열거 상수 명칭 수정(중복 의미 제거)

sinb57 committed:
refactor: condition 클래스의 의미분명한 사용을 위해 enum 제거

sinb57 commented:
다시 한번 더 PR 리뷰 요청드립니다!! 코멘트 달아주신 여러 피드백이 정말 많이 도움되고 있습니다!! 정말 감사합니다!

ksy90101 commented on src/main/java/racingcar/domain/car/condition/NameLength.java:
```diff
@@ -0,0 +1,21 @@
+
+    public static int getMaximum() {
+        return MAXIMUM.number;
+    }
```
굳이 가져다 쓰지 않을 이유가 있나요?
전 Enum에 필드를 만들어서 외부에서 사용할 수 있다는게 Enum의 가장 큰 장점이라고 생각이 듭니다 :) 
혹시 지금까지 이프는 어떻게 Enum 필드를 사용하셨을까요? ㅎㅎ
이건 개인적으로 궁금해서 여쭤봅니다 ㅎㅎ

ksy90101 commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,72 @@
+        this.outputView = outputView;
+    }
+
+    public void initGame() {
```
여기도 한번 테스트 코드를 작성해볼까요?
실제 전체적으로 잘 되는지 어떻게 확인할 수 있을까요?
이프는 충분히 할 수 있을거 같아서 어려운 테스트 이지만, 한번 시도 해보시면 좋을거 같아요!

ksy90101 commented on src/main/java/racingcar/domain/round/Round.java:
```diff
@@ -0,0 +1,22 @@
+        this.count = count;
+    }
+
+    public boolean isNotFinished() {
+        return this.count != 0;
+    }
+
+    public void decreaseCount() {
+        this.count--;
+    }
```
저는 개인적으로 이러한 간단한 메서드도 테스트 코드를 작성해둬야 한다고 생각이 듭니다.
혹시라도 다른 개발자가 정책과 다르게 변경하게 되었을때 유연하지 못할수도 있어서요.

ksy90101 commented on src/main/java/racingcar/service/GameService.java:
```diff
@@ -0,0 +1,49 @@
+import racingcar.domain.round.Round;
+import racingcar.service.picker.NumberPicker;
+
+public class GameService {
```
이쪽도 충분히 테스트 코드를 작성할 수 있을거 같아요.
한번 작성해보실까요?

ksy90101 commented on src/main/java/racingcar/service/picker/RandomNumberPicker.java:
```diff
@@ -0,0 +1,19 @@
+        random.setSeed(System.currentTimeMillis());
+    }
+
+    @Override
+    public int pickNumber() {
+        return RandomRange.pick(random);
+    }
```
여기도 하나의 정책이기 때문에 테스트 코드를 작성해두면 좋을거 같아요.
Junit에서는 isBetween()이러한 메서드를 제공해주고 있습니다.

ksy90101 commented on src/main/java/racingcar/view/output/message/OutputMessage.java:
```diff
@@ -0,0 +1,10 @@
+
+public class OutputMessage {
+
+    public static final String MessageOfRequestCarNames = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+    public static final String MessageOfRequestRound = "시도할 회수는 몇회인가요?";
+    public static final String MessageOfStatusTitle = "실행 결과";
+    public static final String MessageFormatOfPrintWinner = "%s가 최종 우승했습니다.";
```
상수의 명명 컨벤션을 무엇일까요!?

ksy90101 commented on src/main/java/racingcar/view/output/message/OutputMessage.java:
```diff
@@ -0,0 +1,10 @@
+package racingcar.view.output.message;
+
+public class OutputMessage {
```
여기도 생성이 되면 안될거 같은데요.
생성자를 private으로 설정해두면 어떨까요?
아니면 ExceptionMessage때 처럼 enum으로 해주셔도 되구요 ㅎㅎ

ksy90101 commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,43 @@
+    private static final Pattern CUSTOM_SEPARATOR_PATTERN = Pattern.compile("//(.*)\n(.*)");
+    private static final String STANDARD_UNITS = ",:";
+    private static final String SEPARATOR_FORMAT = "[%s]";
+    private static final String emptyString = "";
```
위의 상수와 컨벤션이 다르네요 ㅎㅎ

ksy90101 reviewed (approved):
안녕하세요 이프 :)  리뷰어 럿고 입니다.
리뷰를 남긴것들을 모두 잘 적용해주시고 깊은 고민후에 답변을 남겨주셔서 저도 많이 배웠습니다 💯 
1단계 미션은 여기까지 하도록 하겠습니다. 너무 고생 많으셨어요.

테스트 코드 추가 피드백 및 간단한 리뷰를 남겼는데, 해당 부분은 충분히 2단계에서 또 수정하시면서 적용이 될거라고 생각이 들어서, 머지하도록 하겠습니다.

2단계때 다시 뵙겠습니다!

verus-j commented:
@sinb57 
왜 레이어드 아키텍처를 사용하셨나요?
